### DISCUSSION TITLE
Recovery authentication codes

### DISCUSSION THREAD

**Discussion by rh-iatanaso** [keycloak/keycloak#8518] [Ideas]
## Support for recovery authentication codes

* Status: **Final # 1**
* [https://issues.redhat.com/browse/KEYCLOAK-6270](https://issues.redhat.com/browse/KEYCLOAK-6270)

## Motivation

The base work around multi-factor authentication is defined in [multi-factor-admin-and-step-up.md](https://github.com/keycloak/keycloak-community/blob/master/design/multi-factor-admin-and-step-up.md). However users need an easy alternative means of authentication if they are unable to generate a one time token. Currently, users with 2FA enabled who do not have access to the device configured with an authenticator application must reset their password, which disables 2FA and allows the user to authenticate. In the event that the user’s email account has been compromised, this renders 2FA useless as an attacker can reset the password, disabling 2FA, allowing access to the user’s account. 

We would like to  propose a solution which will deliver a functionality to generate recovery authentication codes that can be printed or stored in a safe location. Those recovery authentication codes can then be used when none of the other forms of second factor authentications are available, or potentially they can be used as a second factor authentication on their own. Quite a few of the authentication providers and websites have features that deliver similar functionality.

## Use Cases

* Allow users to use recovery codes as recovery for 2FA
* Allow users to user recovery code as primary 2FA

## Design Proposal

Recovery codes will be configured as part of the Authentication flow. 

Following requirements will be followed:

* recovery codes will uppercase + digits characters only without 0,O for better human readability 
* Each code will be 3 groups of 4 characters separated by dash
    * Example:  F7IR-5R6N-3RV1 
* Total of 12 recovery authentication codes will be generated at a time
* Each of the recovery code is of single use, hashed with SHA256 
* When new set is generated the old set of recovery codes will be invalidated and removed
* User should be able to print or download the recovery code after they are generated
* Users will be able to remove their recovery codes from the ui with or without generating a new set of recovery codes
* User should be prompted to supply the next unused recovery code
* When user is using his last recovery code he will be prompted to generate new codes.
* Brute force protection should be enabled

## Implementation details

In order to enable the recovery codes the server administrator will register  the ‘Required Action’ through the Realm > Authentication -> Required Action and enable the feature.

Information about recovery authentication codes will be available when managing users in the realm at **Realm > Manage -> Users -> User -> Credentials**. There will be a specific credential type. The number of remaining recovery codes will be available in the Data section.

The user will be able to create or remove recovery codes under the Two-Factor authentication section in the account console. The section will show the number of recovery codes available as well as the date that the recovery codes were generated.  Notification messages will be displayed if the number of recovery codes left is under 3.

When a user is prompted to authenticate the recovery authentication codes should be available as an option if  already has another 2FA method configured. If OTP is disabled the user will be able to authenticate only with recovery authentication codes.Specific recovery code number will be requested. Those would be requested in sequential order starting from 1.

### UI Screens

For simplicity few screens can be found below.

### Account console

* Before setup
<img src="https://user-images.githubusercontent.com/76496192/135494196-9a0663e6-b397-43eb-a90b-3e436916268f.png" width="70%"> 

* Recovery codes display and download screen. User shall be able to copy, download or print the recovery codes
<img src="https://user-images.githubusercontent.com/76496192/145829707-35ead34b-02a1-4839-891a-fefcca6fbbea.png" width="50%">

* Print Preview
<img src="https://user-images.githubusercontent.com/76496192/145829839-5e1af1f9-ef24-4d11-b146-982a5ac37a57.png" width="50%">

* After setup - the number of the remaining recovery codes will be shown on this screen.
<img src="https://user-images.githubusercontent.com/76496192/139459473-50d21a0f-63f8-4f77-b371-c0bf7f109092.png" width="70%"> 

* Login screen authenticator selector 
<img src="https://user-images.githubusercontent.com/76496192/138278535-e46f31b3-72df-442b-8c1d-e72cd587767c.png" width="50%">

* Login screen recovery code request
<img src="https://user-images.githubusercontent.com/76496192/135494204-30dac6c6-9776-461e-8fe6-ca0b96f8a64c.png" width="50%">

### Admin Console
* Authentication Flow configuration
<img src="https://user-images.githubusercontent.com/76496192/135498967-4d6f4f50-21ae-47c8-98ec-a01c59aa94f9.png" width="70%">

### Admin console  >> User management

* Credentials management - the remaining number of codes can be found in the Data section
<img src="https://user-images.githubusercontent.com/76496192/135499244-0ff52f3a-29c6-4700-8abe-c96d43376af7.png" width="70%">

### Database changes

There will be no database changes required. We are using the CredentialModel to store the credentials.

## Resources

* This discussion was influenced by Thomas Darimont work at https://github.com/thomasdarimont/keycloak-extension-playground/tree/master/auth-backup-codes
* Guidance from from https://cheatsheetseries.owasp.org/cheatsheets/Forgot_Password_Cheat_Sheet.html#backup-codes 

### Recovery Codes entropy 

* 12 characters gives the following entropy:

Character class | entropy | permutations
----------------------|-----------|---------------------
digits:		 |	39.9 |   10<sup>12</sup> 
**upper+digits-(O,0)** |      61 |  2.39×10<sup>18</sup> 
upper+digits:	 |	62.0  | 4.73×10<sup>18</sup> 
lower+upper: 	 |	68.0  | 3.9×10²⁰ 
lower+upper+digits: | 71.0 | 3.2×10²¹ 

Comparing to others:
Provider | Character class | entropy | permutations
------------|----------------------|------------|----------------
Google | 8 digits |	26.5 | 10<sup>8</sup> 
GitHub |10 lower+digits|  51.7  |   4.7×10¹⁸
Facebook |8 digits |	 26.5 | 10<sup>8</sup> 
Twitter (single backup code) | 12 lower+digits | 62.0 |  4.7×10¹⁸

**Comment by stianst** [comment]
Milestones should probably be removed.

I presume "Recovery codes display and download screen" isn't the final product?

**Reply by rh-iatanaso** [reply]
I made an update. There will be download link , also when displayed the codes will be probably split in 3 columns. We will display also the remaining number codes in the account console

**Reply by stianst** [reply]
@rh-iatanaso you mean the remaining number of codes right, not the actual codes themselves?

**Reply by rh-iatanaso** [reply]
Thanks and yes  that is what I actually reflected in the discussion. I updated in my comment as well.

**Reply by raphaeldemedion** [reply]
the process is ........

**Comment by jbman** [comment]
I like that this proposal contains a detailed description and screenshots. Makes it easier to think through and provide feedback 👍 

My notes on UI/UX:
- If the user has no recovery codes created, a screen asks if it should be done now, right? If all codes are used, the same should happen 
- The user should be advised to store the backup codes not at the authentication device but at an other place 
- If user can remove backup codes it should be possible to find the stored/printed codes. The creation date and time should be displayed at both places: after backup code creation and where they can be removed.
- Number of remaining codes is not shown in the screenshot, also described in the text.
- If he codes are called recovery codes and not intended for regular use, I don't see why you need 15 of them at a time. 10 or only 5 should do. Also doesn't need multi culumn display.
- Right after after a successful ogin with recovery code, the user should be asked if a new second factor should be replace and invalidate the old one.
- Instead of numbered codes it could also be nice if any code can be tried (e.g 5 codes equals 5 tries allowed). Even OWASP says 
> A user should have multiple recovery codes at any given time to ensure that one of them works

**Reply by rh-iatanaso** [reply]
Thank you for your feedback.
Please find my comments below:

* On the first point recovery codes are associated with 2FA  so if user has OTP enabled and a device configured we can ask them to setup recovery codes. When 1 or 2 recovery codes are left user will be advised to generate new ones. 
* Yes we will can add more informative description to educate the users
* We will add data generated in the printout  , the date is already shown in the account console.
* Yes this is gap in the screenshot and pending the outcome of this discussion
* The codes are called recovery but they can be used regularly as well . That is why we think 15 codes is a good number . Ultimately once initial implementation is done we can add configuration for the number of codes to be generated between min and max. Having multiple columns makes it a bit tidy in my opinion but that is up for this discussion.
*   I am not sure if we should be asking the person to replace his 2FA could be just that device was not available at this moment and the recovery codes are 
* We want the ask user for specific code because that makes it easy to track which one was used and which one was not by the user. Otherwise the user _must_ keep track of used recovery codes. We can probably ask them of order that might safe the user if they only have half the codes by some reason. We do want to avoid user guessing because if brute force protection is enable this might disable his/her account.

**Reply by rh-iatanaso** [reply]
@jbman in the owasp reference I added, the suggested number was up to 10 but, the number is really arbitrary in my opinion probably anything between 5 and 30 would make sense for me at least. As  I mentioned we can make it configuration if there is need for it.

**Comment by thomasmicro** [comment]
Thanks for the thoughts and the PoC!! I really like the feature!

I agree with @jbman that the user should be able to enter any of the backup-codes. Maybe a way to know which code the user entered could be a list like this:
- 00-F7IO-5O6N-3OV1
- 01-YW2S-YT58-GMYB
- 02-M0UJ-ENEM-DJH0
- 03-DH12-PKPZ-6652
- 04-GG2P-J6RX-R8CF
- 05-6KGT-HM1Q-P78L
- 06-2PQH-HQ7G-FLU0
- 07-8XEJ-UJ1Z-XYJZ
- ...

And then extracting the first two digits. This would increase the length of a backup code . Or we do it like in https://groups.google.com/g/keycloak-dev/c/SC1JvewgLwM/m/pIAS4-9pAwAJ and just try every code with a reduced number of iterations.

Thanks for your work!
Thomas

**Reply by stianst** [reply]
I don't agree with having the option to enter any backup codes as it gives both a worse user experience as it's harder for the user to know what codes have been used and how many there are left, and also reduced security as an attacker can guess from many codes rather than a single one. It also is more costly (although that's insignificant really) as rather than compare a single code you need to compare multiple codes.

**Reply by jbman** [reply]
When searching for "2FA recovery codes" with image search, it seems to best practice to provide multiple valid codes without enforcing linear use. Github, Google and Facebook do it like that. I also wonder why this is the established way. Of cause you can split your codes and store them at different places. You can hand one out to a trusted recovery person. You can print them and take all except one with you, instead of your phone. 

Many providers also allow to view the recovery codes when you're logged in (e.g. Github and Facebook do). Github tells me that my codes were "viewed X hours ago".  If one time codes are used as 2FA replacement, this seems to be common practice. Maybe usability was preferred over best possible security. 

Github also brings me right to the 2FA settings, if I log in with a recovery code: 
> Having trouble with two-factor authentication? You can update your settings here

**Comment by abstractj** [comment]
@rh-iatanaso thanks for putting this together, very concise and easy to understand proposal. I'm curious about the statement below:

"_When a user is prompted to authenticate the recovery authentication codes should be available as an option if already has another 2FA method configured. If OTP is disabled the user will be able to authenticate only with recovery authentication codes_"

Let's think about the following malicious user story:
```
As an attacker

I want to steal and use the recovery codes
so that I can lock the user outside, by disabling the OTP
```

Please correct me if I'm wrong, but it seems that the adversary should be able to log in with the recovery authentication code, disable OTP and lock the user outside the app. Is that correct?

**Reply by stianst** [reply]
If an adversary has the password and recovery code that is correct. Recover codes are an alternative 2nd factor.

There's not really any other way to do it as the legitimate user also nerds to be able to remove an old OTP and add a new one.

**Reply by stianst** [reply]
What can and should be done is to start thinking about adaptive/risk/continuous verification that is an extra layer of defence around authentication.

**Reply by thomasmicro** [reply]
You mean the attacker finds a browser window with a logged in user and disabled OTP to lock the user out until the user gets the recovery keys? I think the "disable OTP" should be a AIA requiring Authentication. Also we could attach the recovery keys to certain 2FA-Token and delete them with the 2FA-Token (thats differing from the proposal though).

**Reply by stianst** [reply]
For sure disable OTP should be an AIA. Isn't that already the case?

**Reply by thomasmicro** [reply]
No, clicking the button in the account-console just did a DELETE /auth/realms/Test/account/credentials/c580adf5-7a1f-4ece-b21c-a1db9da481ac for me.

**Reply by rh-iatanaso** [reply]
Should we make recovery codes deletion AIA  as well ?

**Reply by abstractj** [reply]
@rh-iatanaso I'd suggest AIA for disabling OTP, plus the action of deleting the recovery codes.

**Reply by rh-iatanaso** [reply]
@abstractj  Considering using AIA for disabling OTP will be change of prior behavior should it be separate issue ?

**Reply by stianst** [reply]
For sure AIA for disabling OTP doesn't really have anything to do with recovery codes. So should be opened as a JIRA / bug.

**Reply by thomasmicro** [reply]
I just opened one - feel free to edit it: https://issues.redhat.com/browse/KEYCLOAK-19673

**Reply by rh-iatanaso** [reply]
After looking into it some more , currently  only Create/Update actions support AIA. CredentialTypeMetadata does not have support for update action. I think the issue above should handle the generic case to add support for Delete actions to use AIA.
At this moment enabling this AIA for recovery codes will be a hack . We think this should be addressed in general code and add AIA for Delete

**Comment by DGuhr** [comment]
Great work and great proposal so far, thanks for that! 

I just try to get my head around the "Allow users to use recovery code as primary 2FA" - Part. Could you clarify that a bit?

_disclaimer: The following is more a "stream of conciousness" with a few questions inside from my side w.r.t. using the codes as primary 2FA:_

I imagine a user has not setup 2FA yet, but creates recovery codes (thus, leading to 2FA enabled for the login flow).
Then he/she has essentially x Logins left, before... what happens? 

potential options of "what happens" i think of:

option 1: user gets noticed to re-create new codes (or setup an MFA device) before the codes are completely gone?
pro: this, perhaps together with configurable number of recovery codes, could open up pretty interesting "static mfa" use cases, like creating a dedicated realm where you have only one recoverycode without possibility to re-create => kind of rate-limiting to one-time login for people.
con: Then the name may be misleading imo. "recovery" pretty much settles the use case.

option 2: user gets no notice, and has to reset his/her account after using the 15 codes.
pro: nothing I could think of.
con: the process itself. (But: Is it possible to disable the future usage of recovery codes after successfull login?)

option 3: make the usage of recovery codes dependant on having set up a 2FA Generator/APP at least once
pro: "recovery" makes sense in my head now ;)
con: essentially that would be a removal of the "feature" to be able to use these codes as primary 2FA.

So, as said, great work so far. Would be great if you could clarify the question :)

**Reply by rh-iatanaso** [reply]
Thanks for the input I really appreciate it.
Our thinking is more aligned with option 1 with small differences.
User will be prompted to generate recovery codes when he exhausts them and warned that failing to do so will impact his ability to login. We will also  warn the user with account console notification where the codes are low in number.
I agree with observation that "recovery" might be misleading in the case where user is not using other 2FA generator. 
Option 3: We can detect if 2FA is setup, but there might be a case which user had requested his 2FA generator to be removed for his account. Then we are back in option 1. However maybe we should put emphasis in the UI and documentation that the primary reason to use recovery code is for recovery not for separate 2FA .

**Reply by SecKatie** [reply]
I am concerned about using recovery codes as the primary 2FA mechanism, and I understand that it is a secondary use case, but it is something that I would say has very little value. The purpose of standard 2FA is "something you have and something you know," with the password being something you know. Recovery codes are explicitly made to be written down so that they become "something you have."

I could see a valid use case for recovery codes if a user doesn't have a smartphone or physical token. When I used to work at an IT helpdesk, some users would not be able to utilize our recommended 2FA mechanisms due to this, and printed recovery codes might work for them.

My concern occurs when a user is on an untrusted computer, like at a library or on a school PC, and they are unable to print or download the codes to a valid location (if they do not have a smartphone, they likely will not be able to capture a picture either). If this occurs on their last login they would be locked out.

From my perspective, if the last 2FA generator or hardware token is removed, then the recovery codes should be as well.

**Reply by rh-iatanaso** [reply]
* I think user can always use pen and paper if nothing else is available.
* Part of the reason we would like to have recovery codes is exactly in case the 2FA is removed , let's say lost phone or hardware token, user still has second option and not depending only on password.

Ultimately this is another tool in the tool box that will help administrators to provide best experience for their users.

**Reply by SecKatie** [reply]
> * Part of the reason we would like to have recovery codes is exactly in case the 2FA is removed , let's say lost phone or hardware token, user still has second option and not depending only on password.

For this case, it is necessary to make it clear that Recovery Codes are not removed when 2FA is removed or give the option for the user to disable both at the same time. I am not aware of a service that leaves Recovery Codes enabled when 2FA is disabled so proper user education will be key to preventing problems.

**Reply by rh-iatanaso** [reply]
I think Facebook and Twitter have on/off switch for 2FA which is more explicit, keycloak does not have on/off switch. But if the consensus is that we want to remove the backup codes if all the 2FA methods are deleted we can do that. But in general recovery codes + password will be better than just password if I just lost my phone and I need some time to procure new device.

**Reply by stianst** [reply]
Keycloak doesn't currently have an option to "disable 2FA", but rather that is implicit when all 2FA authenticators are removed. I do agree that a switch to "enable/disable" 2FA could be useful, but that should be discussed as a separate thing to introducing recovery codes.

**Comment by mposolda** [comment]
Few more points from me besides what was already told here:
- I think it will be nice if required action is registered by default? It is one less step for administrator, which IMO is fine.
- I am not 100% sure if we want both OTP and Recovery codes to be available in default authentication flow by default? Or maybe we want to do add it later as a follow-up work? (Possibly also in relation to authentication policies?)
- It will be nice to have icon on the "Login screen authenticator selector" , but I understand it is likely not the biggest priority :)
- My vote is that Keycloak will ask user about concrete recovery code. For example: "Please use recovery code number 7" instead of rely on user to use any random code. I am not seeing any advantage of using random code by the user? I can see just disadvantages, which were already mentioned...
- It will be nice that after using of all the 15 codes during authentication, Keycloak will display something after authentication with last code was done. Something like "All your recovery codes were used. Do you want to setup new codes now?" . If user press YES, he can setup new codes right away during authentication. Otherwise he can still use AIA to be triggered from account console and do it later

**Reply by thomasmicro** [reply]
From your point-of-view: Would it be possible for users to set up multiple independent sets of recovery codes, so people can choose the set they have available at their current location?

I would definitly warn the users in the UI that they have to have ALL recovery codes of a set available to securely recover a account - since I would assume the opposite.

**Reply by rh-iatanaso** [reply]
@mposolda 
* if everyone agrees we can make the required action to be registered by default
* I don't think we want to add OTP and Recovery codes to be available in default action for now. This might be confusing for the end user.
* There will be an icon for sure and it will be part of our initial PR
*  This is some what at tie we can start with numbered codes and then switch to random, the other might be bit confusing for the end user  to switch to later.
* There will be warning and notifications. 
@thomasmicro We are currently not proposing or planning for multiple independent sets of recovery codes.

**Comment by stianst** [comment]
![image](https://user-images.githubusercontent.com/2271511/139638240-69d5602d-fa8c-409c-a8c8-ab455a76f6dc.png)

I noticed when viewing the codes one of the codes is wrapped on multiple lines, which is not great. Perhaps we should go with 2 columns instead of 3, then go down to 12 codes?

**Reply by rh-iatanaso** [reply]
In general number that divides but 2 and 3 will be better as starting point because it provides more flexibility you can have 2,3,4 columns .
I like the idea of having 12 codes. I am not sure wrapping can be prevented 100% because it is partial based on user screen resolution

**Reply by stianst** [reply]
Can we go down to a single column if the width is not wide enough for the codes to be displayed on a single line? Would look quite daft if at lower res we have two columns and at the same time have codes spanning two rows

**Reply by asumihasan** [reply]
In my designs for the next iteration, I have the numbered list being separate from the recovery codes so that the numbers can be right-aligned. This helps when the column contains both single and double-digit numbers (ex. 9 and 10 in the screenshot) and should prevent the beginning of the recovery code from being pushed to the right and consequently wrapping. Not sure how this impacts development etc., but this is just from the visual standpoint.

**Comment by thomasdarimont** [comment]
This looks good and is much more idiomatic than my initial implementation :)

Some remarks about how we currently use the original backup code implementation:
- We call those codes "Backup Codes" instead of "Authentication Recovery Codes"
- The backup codes are only an alternative to a configured 2nd factor, e.g. OTP, SMS, Web Push
- If the user removed the last 2nd factor the backup codes are also removed
- A user can only login with a backup code if a 2nd factor is registered for the user
- We prompt the user to generate and download backup codes after the registration of a 2nd factor, iif no backup codes are present yet. Users can decide to postpone the backup-code generation action, if the user decides so, then a hint for generating backup-codes is shown in the custom account-console.
- If we detect that a user has registered a 2nd factor but did not create backup codes after 7 days, we'll send a friendly reminder email.
- In our current implementation, the user can enter any unused backup code - the user needs to keep track of which code is already used.

Regarding the generated codes:
- The generated codes might contain ambiguous characters like 0 (zero) and O (uppercase 0). How about generating those numbers with a special alphabet that avoids those combinations, like [crockford-base-32-encoding](https://www.dcode.fr/crockford-base-32-encoding)? This helps to avoid typos and might reduce support requests.

**Reply by rh-iatanaso** [reply]
Thomas,
Thank you for your reply.
* Currently the name we are using Recovery Authentication Codes for us it just sounded correct. 
* I think the consensus is so far we should remove the codes when user removes all his 2nd factor.
* same as above
* same as above
* I think generate recovery code is clearly marked and the users that want to do that will probably do. For non technical user it will be bit confusing to they they need recovery codes. I would deffer that as something that could be done later if there is need for it.
* I am not sure if everyone will want to send email or even get email with that reminder.
* Tracking the backup codes for the user makes it easy if we you have brute force protection enabled and set on low threshold .

We picked uppercase + numbers just to avoid that problem  I (uppercase i ) , l ( lowercase L) , crockford-base-32-encoding might good to avoid the O(0) problem .

**Comment by stelewis-redhat** [comment]
The round of changes after review that affects the features include:
1. Codes are updatable within the account console. Instead of removing then re-creating.
2. Direct grant feature has been removed.
3. Help text is now "These codes can be used to regain your access in case your other 2FA means are not available"
4. There is now a feature flag to enable recovery codes. It's currently a preview feature.

Everything else is a refactoring.

See https://github.com/keycloak/keycloak/pull/8730 for full discussion.
